src/backend/optimizer/plan/transform.c

```diff
@@ -520,3 +521,124 @@ replace_sirvf_rte(Query *query, RangeTblEntry *rte)
 
 	return rte;
 }
+
+/*
+ * Does query has SRFs, or WITH ORDINALITY?
+ */
+bool query_has_srf(Query *query)
+{
+	if (query->hasTargetSRFs)
+	{
+		return true;
+	}
+
+	/* Double check for subquery. */
+	if (expression_returns_set( (Node *) query->targetList))
+	{
+		return true;
+	}
+
+	ListCell *lcrte = NULL;
+	foreach (lcrte, query->rtable)
```

Comment:
Do we need to check the rtable? If the rtable contains any srf function, I think that's ok, just as a relation which has multiple rows.

```
 explain select distinct(count(a)) from generate_series(0,10000) as a;
                                          QUERY PLAN
----------------------------------------------------------------------------------------------
 Unique  (cost=125.04..125.04 rows=1 width=8)
   Group Key: (count(a))
   ->  Sort  (cost=125.04..125.04 rows=1 width=8)
         Sort Key: (count(a))
         ->  Aggregate  (cost=125.02..125.03 rows=1 width=8)
               ->  Function Scan on generate_series a  (cost=0.00..100.01 rows=10001 width=4)
 Optimizer: Postgres query optimizer
(7 rows)
```
In the above case, the distinct expression should be optimized too.

Reply:
>  sql explain select distinct(count(a)) from generate_series(0,10000) as a;
>  In the above case, the distinct expression should be optimized too.

You're right.